File: ban.go
package main

import (
	"log"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api"
)

func ban(bot *tgbotapi.BotAPI, update tgbotapi.Update) {
	if update.Message.ReplyToMessage == nil {
		bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
		msg := tgbotapi.NewMessage(update.Message.Chat.ID, "Uhm? No puedo banear a nadie si no respondes a un mensaje")
		msg.ReplyToMessageID = update.Message.MessageID
		bot.Send(msg)
		return
	}

	chatMember, err := bot.GetChatMember(tgbotapi.ChatConfigWithUser{
		ChatID: update.Message.Chat.ID,
		UserID: update.Message.From.ID,
	})
	if err != nil {
		log.Println(err)
		return
	}

	if !chatMember.CanRestrictMembers && chatMember.Status != "creator" {
		bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
		msg := tgbotapi.NewMessage(update.Message.Chat.ID,
			"Lo siento, pero me dijeron que tú no puedes banear usuarios...")
		msg.ReplyToMessageID = update.Message.MessageID
		bot.Send(msg)
		return
	}

	chatMember, err = bot.GetChatMember(tgbotapi.ChatConfigWithUser{
		ChatID: update.Message.Chat.ID,
		UserID: bot.Self.ID,
	})
	if err != nil {
		log.Println(err)
		return
	}

	if !chatMember.CanRestrictMembers {
		bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
		msg := tgbotapi.NewMessage(update.Message.Chat.ID, "Etto... no tengo los permisos para banear usuarios (╥_╥)")
		msg.ReplyToMessageID = update.Message.MessageID
		bot.Send(msg)
		return
	}

	var user *tgbotapi.User
	if update.Message.ReplyToMessage.ForwardFrom != nil {
		user = update.Message.ReplyToMessage.ForwardFrom
	} else {
		user = update.Message.ReplyToMessage.From
	}

	if user.ID == bot.Self.ID {
		bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
		msg := tgbotapi.NewMessage(update.Message.Chat.ID, "Hey! Esa soy yo! (O.O)")
		msg.ReplyToMessageID = update.Message.MessageID
		bot.Send(msg)
		bot.Send(tgbotapi.NewStickerShare(update.Message.Chat.ID, "CAADAgADrRsAAuCjgge1g6be76IiHgI"))
	}

	if update.Message.ReplyToMessage.ForwardFrom != nil {
		chatMember, err = bot.GetChatMember(tgbotapi.ChatConfigWithUser{
			ChatID: update.Message.Chat.ID,
			UserID: update.Message.ReplyToMessage.ForwardFrom.ID,
		})
		if err != nil {
			log.Println(err)
			return
		}
	} else {
		chatMember, err = bot.GetChatMember(tgbotapi.ChatConfigWithUser{
			ChatID: update.Message.Chat.ID,
			UserID: update.Message.ReplyToMessage.From.ID,
		})
		if err != nil {
			log.Println(err)
			return
		}
	}

	if chatMember.Status == "administrator" || chatMember.Status == "creator" {
		bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
		msg := tgbotapi.NewMessage(update.Message.Chat.ID, "Etto... No puedo banear administradores (⌒_⌒;)")
		msg.ReplyToMessageID = update.Message.MessageID
		bot.Send(msg)
		return
	}

	if update.Message.ReplyToMessage.ForwardFrom != nil {
		bot.KickChatMember(tgbotapi.KickChatMemberConfig{
			ChatMemberConfig: tgbotapi.ChatMemberConfig{
				ChatID: update.Message.Chat.ID,
				UserID: update.Message.ForwardFrom.ID,
			},
		})
	} else {
		bot.KickChatMember(tgbotapi.KickChatMemberConfig{
			ChatMemberConfig: tgbotapi.ChatMemberConfig{
				ChatID: update.Message.Chat.ID,
				UserID: update.Message.ReplyToMessage.From.ID,
			},
		})
	}

	bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
	msg := tgbotapi.NewMessage(update.Message.Chat.ID, "¡Banead@!")
	msg.ReplyToMessageID = update.Message.MessageID
	bot.Send(tgbotapi.NewStickerShare(update.Message.Chat.ID, "CAADAgADpxsAAuCjggf0_Fu4xLzgxAI"))
}

func kick(bot *tgbotapi.BotAPI, update tgbotapi.Update) {
	if update.Message.ReplyToMessage == nil {
		bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
		msg := tgbotapi.NewMessage(update.Message.Chat.ID,
			"Uhm? No puedo expulsar a nadie si no respondes a un mensaje")
		msg.ReplyToMessageID = update.Message.MessageID
		bot.Send(msg)
		return
	}

	chatMember, err := bot.GetChatMember(tgbotapi.ChatConfigWithUser{
		ChatID: update.Message.Chat.ID,
		UserID: update.Message.From.ID,
	})
	if err != nil {
		log.Println(err)
		return
	}

	if !chatMember.CanRestrictMembers && chatMember.Status != "creator" {
		bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
		msg := tgbotapi.NewMessage(update.Message.Chat.ID,
			"Lo siento, pero me dijeron que tú no puedes expulsar usuarios...")
		msg.ReplyToMessageID = update.Message.MessageID
		bot.Send(msg)
		return
	}

	chatMember, err = bot.GetChatMember(tgbotapi.ChatConfigWithUser{
		ChatID: update.Message.Chat.ID,
		UserID: bot.Self.ID,
	})
	if err != nil {
		log.Println(err)
		return
	}

	if !chatMember.CanRestrictMembers {
		bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
		msg := tgbotapi.NewMessage(update.Message.Chat.ID, "Etto... no tengo los permisos para expulsar usuarios (╥_╥)")
		msg.ReplyToMessageID = update.Message.MessageID
		bot.Send(msg)
		return
	}

	var user *tgbotapi.User
	if update.Message.ReplyToMessage.ForwardFrom != nil {
		user = update.Message.ReplyToMessage.ForwardFrom
	} else {
		user = update.Message.ReplyToMessage.From
	}

	if user.ID == bot.Self.ID {
		bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
		msg := tgbotapi.NewMessage(update.Message.Chat.ID, "Hey! Esa soy yo! (O.O)")
		msg.ReplyToMessageID = update.Message.MessageID
		bot.Send(tgbotapi.NewStickerShare(update.Message.Chat.ID, "CAADAgADrRsAAuCjgge1g6be76IiHgI"))
		return
	}

	if update.Message.ReplyToMessage.ForwardFrom != nil {
		chatMember, err = bot.GetChatMember(tgbotapi.ChatConfigWithUser{
			ChatID: update.Message.Chat.ID,
			UserID: update.Message.ReplyToMessage.ForwardFrom.ID,
		})
		if err != nil {
			log.Println(err)
			return
		}
	} else {
		chatMember, err = bot.GetChatMember(tgbotapi.ChatConfigWithUser{
			ChatID: update.Message.Chat.ID,
			UserID: update.Message.ReplyToMessage.From.ID,
		})
		if err != nil {
			log.Println(err)
			return
		}
	}

	if chatMember.Status == "administrator" || chatMember.Status == "creator" {
		bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
		msg := tgbotapi.NewMessage(update.Message.Chat.ID, "Etto... No puedo expulsar administradores (⌒_⌒;)")
		msg.ReplyToMessageID = update.Message.MessageID
		bot.Send(msg)
		return
	}

	if update.Message.ReplyToMessage.ForwardFrom != nil {
		bot.KickChatMember(tgbotapi.KickChatMemberConfig{
			ChatMemberConfig: tgbotapi.ChatMemberConfig{
				ChatID: update.Message.Chat.ID,
				UserID: update.Message.ReplyToMessage.ForwardFrom.ID,
			},
		})
		bot.UnbanChatMember(tgbotapi.ChatMemberConfig{
			ChatID: update.Message.Chat.ID,
			UserID: update.Message.ReplyToMessage.ForwardFrom.ID,
		})
	} else {
		bot.KickChatMember(tgbotapi.KickChatMemberConfig{
			ChatMemberConfig: tgbotapi.ChatMemberConfig{
				ChatID: update.Message.Chat.ID,
				UserID: update.Message.ReplyToMessage.From.ID,
			},
		})
		bot.UnbanChatMember(tgbotapi.ChatMemberConfig{
			ChatID: update.Message.Chat.ID,
			UserID: update.Message.ReplyToMessage.From.ID,
		})
	}

	bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
	msg := tgbotapi.NewMessage(update.Message.Chat.ID, "¡Expulsad@!")
	msg.ReplyToMessageID = update.Message.ReplyToMessage.MessageID
	bot.Send(msg)
	bot.Send(tgbotapi.NewStickerShare(update.Message.Chat.ID, "CAADAgADpxsAAuCjggf0_Fu4xLzgxAI"))
}

func kickMe(bot *tgbotapi.BotAPI, update tgbotapi.Update) {
	chatMember, err := bot.GetChatMember(tgbotapi.ChatConfigWithUser{
		ChatID: update.Message.Chat.ID,
		UserID: bot.Self.ID,
	})
	if err != nil {
		log.Println(err)
		return
	}

	if !chatMember.CanRestrictMembers {
		bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
		msg := tgbotapi.NewMessage(update.Message.Chat.ID, "Etto... no tengo los permisos para expulsar usuario (╥_╥)")
		msg.ReplyToMessageID = update.Message.MessageID
		bot.Send(msg)
		return
	}

	chatMember, err = bot.GetChatMember(tgbotapi.ChatConfigWithUser{
		ChatID: update.Message.Chat.ID,
		UserID: update.Message.From.ID,
	})
	if err != nil {
		log.Println(err)
		return
	}

	if chatMember.Status == "administrator" || chatMember.Status == "creator" {
		bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
		msg := tgbotapi.NewMessage(update.Message.Chat.ID, "Etto... No puedo expulsar administradores (⌒_⌒;)")
		msg.ReplyToMessageID = update.Message.MessageID
		bot.Send(msg)
		return
	}

	bot.KickChatMember(tgbotapi.KickChatMemberConfig{
		ChatMemberConfig: tgbotapi.ChatMemberConfig{
			ChatID: update.Message.Chat.ID,
			UserID: update.Message.From.ID,
		},
	})
	bot.UnbanChatMember(tgbotapi.ChatMemberConfig{
		ChatID: update.Message.Chat.ID,
		UserID: update.Message.From.ID,
	})

	bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
	msg := tgbotapi.NewMessage(update.Message.Chat.ID, "¡Espero que vuelvas pronto! ( ; ω ; )")
	msg.ReplyToMessageID = update.Message.MessageID
	bot.Send(msg)
	bot.Send(tgbotapi.NewStickerShare(update.Message.Chat.ID, "CAADAgADmhsAAuCjggeKLIclx5HvGQI"))
}

func unban(bot *tgbotapi.BotAPI, update tgbotapi.Update) {
	if update.Message.ReplyToMessage != nil {
		bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
		msg := tgbotapi.NewMessage(update.Message.Chat.ID, "Uhm? No puedo banear a nadie si no respondes a un mensaje")
		msg.ReplyToMessageID = update.Message.MessageID
		bot.Send(msg)
		return
	}

	chatMember, err := bot.GetChatMember(tgbotapi.ChatConfigWithUser{
		ChatID: update.Message.Chat.ID,
		UserID: update.Message.From.ID,
	})
	if err != nil {
		log.Println(err)
		return
	}

	if !chatMember.CanRestrictMembers && chatMember.Status != "creator" {
		bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
		msg := tgbotapi.NewMessage(update.Message.Chat.ID,
			"Lo siento, pero me dijeron que tú no puedes desbanear usuarios...")
		msg.ReplyToMessageID = update.Message.MessageID
		bot.Send(msg)
		return
	}

	chatMember, err = bot.GetChatMember(tgbotapi.ChatConfigWithUser{
		ChatID: update.Message.Chat.ID,
		UserID: bot.Self.ID,
	})
	if err != nil {
		log.Println(err)
		return
	}

	if !chatMember.CanRestrictMembers {
		bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
		msg := tgbotapi.NewMessage(update.Message.Chat.ID,
			"Etto... no tengo los permisos para desbanear usuarios (╥_╥)")
		msg.ReplyToMessageID = update.Message.MessageID
		bot.Send(msg)
		return
	}

	if update.Message.ReplyToMessage.ForwardFrom != nil {
		bot.UnbanChatMember(tgbotapi.ChatMemberConfig{
			ChatID: update.Message.Chat.ID,
			UserID: update.Message.ReplyToMessage.ForwardFrom.ID,
		})
	} else {
		bot.UnbanChatMember(tgbotapi.ChatMemberConfig{
			ChatID: update.Message.Chat.ID,
			UserID: update.Message.ReplyToMessage.From.ID,
		})
	}

	bot.Send(tgbotapi.NewChatAction(update.Message.Chat.ID, tgbotapi.ChatTyping))
	msg := tgbotapi.NewMessage(update.Message.Chat.ID, "¡Desbanead@! (o^▽^o)")
	msg.ReplyToMessageID = update.Message.ReplyToMessage.MessageID
	bot.Send(msg)
	bot.Send(tgbotapi.NewStickerShare(update.Message.Chat.ID, "CAADAgADshsAAuCjgge9W_YhLXZXrgI"))
}

File: src/ban.ts
import { Composer } from 'telegraf';
import { User } from 'telegraf/typings/telegram-types';

const Ban = new Composer();

Ban.command('ban', async ctx => {
	if (ctx.message!.reply_to_message == undefined) {
		await ctx.replyWithChatAction('typing');
		await ctx.reply('Uhm? No puedo banear a nadie si no respondes a un mensaje', {
			reply_to_message_id: ctx.message!.message_id
		});
		return;
	}

	let chatMember = await ctx.getChatMember(ctx.from!.id);
	if (!chatMember.can_restrict_members && chatMember.status != 'creator') {
		await ctx.replyWithChatAction('typing');
		await ctx.reply('Lo siento, pero me dijeron que tú no puedes banear usuarios...');
		return;
	}

	chatMember = await ctx.getChatMember(ctx.botInfo?.id ?? (await ctx.tg.getMe()).id);
	if (!chatMember.can_restrict_members) {
		await ctx.replyWithChatAction('typing');
		await ctx.reply('Etto... no tengo los permisos para banear usuarios (╥_╥)', {
			reply_to_message_id: ctx.message!.message_id
		});
		return;
	}

	let user: User;
	if (ctx.message!.reply_to_message.forward_from != undefined) {
		user = ctx.message!.reply_to_message.forward_from;
	} else {
		user = ctx.message!.reply_to_message.from!;
	}

	if (user.id == (ctx.botInfo?.id ?? (await ctx.tg.getMe()).id)) {
		await ctx.replyWithChatAction('typing');
		await ctx.reply('Hey! Esa soy yo! (O.O)', { reply_to_message_id: ctx.message!.message_id });
		await ctx.replyWithSticker('CAADAgADrRsAAuCjgge1g6be76IiHgI');
		return;
	}

	if (ctx.message!.reply_to_message.forward_from != undefined) {
		chatMember = await ctx.getChatMember(ctx.message!.reply_to_message.forward_from.id);
	} else {
		chatMember = await ctx.getChatMember(ctx.message!.reply_to_message.from!.id);
	}

	if (chatMember.status == 'administrator' || chatMember.status == 'creator') {
		await ctx.replyWithChatAction('typing');
		await ctx.reply('Etto... No puedo banear administradores (⌒_⌒;)', { reply_to_message_id: ctx.message!.message_id });
		return;
	}

	if (ctx.message!.reply_to_message.forward_from != undefined) {
		ctx.kickChatMember(ctx.message!.reply_to_message.forward_from.id);
	} else {
		ctx.kickChatMember(ctx.message!.reply_to_message.from!.id);
	}

	await ctx.replyWithChatAction('typing');
	await ctx.reply('¡Banead@!', { reply_to_message_id: ctx.message!.reply_to_message.message_id });
	await ctx.replyWithSticker('CAADAgADpxsAAuCjggf0_Fu4xLzgxAI');
});

Ban.command('kick', async ctx => {
	if (ctx.message!.reply_to_message == undefined) {
		await ctx.replyWithChatAction('typing');
		await ctx.reply('Uhm? No puedo expulsar a nadie si no respondes a un mensaje', {
			reply_to_message_id: ctx.message!.message_id
		});
		return;
	}

	let chatMember = await ctx.getChatMember(ctx.message!.from!.id);
	if (!chatMember.can_restrict_members && chatMember.status != 'creator') {
		await ctx.replyWithChatAction('typing');
		await ctx.reply('Lo siento, pero me dijeron que tú no puedes expulsar usuarios...', {
			reply_to_message_id: ctx.message!.message_id
		});
		return;
	}

	chatMember = await ctx.getChatMember(ctx.botInfo?.id ?? (await ctx.tg.getMe()).id);
	if (!chatMember.can_restrict_members) {
		await ctx.replyWithChatAction('typing');
		await ctx.reply('Etto... no tengo los permisos para expulsar usuarios (╥_╥)', {
			reply_to_message_id: ctx.message!.message_id
		});
		return;
	}

	let user: User;
	if (ctx.message!.reply_to_message.forward_from != undefined) {
		user = ctx.message!.reply_to_message.forward_from;
	} else {
		user = ctx.message!.reply_to_message.from!;
	}

	if (user.id == (ctx.botInfo?.id ?? (await ctx.tg.getMe()).id)) {
		await ctx.replyWithChatAction('typing');
		await ctx.reply('Hey! Esa soy yo! (O.O)', { reply_to_message_id: ctx.message!.message_id });
		await ctx.replyWithSticker('CAADAgADrRsAAuCjgge1g6be76IiHgI');
		return;
	}

	if (ctx.message!.reply_to_message.forward_from != undefined) {
		chatMember = await ctx.getChatMember(ctx.message!.reply_to_message.forward_from.id);
	} else {
		chatMember = await ctx.getChatMember(ctx.message!.reply_to_message.from!.id);
	}

	if (chatMember.status == 'administrator' || chatMember.status == 'creator') {
		await ctx.replyWithChatAction('typing');
		await ctx.reply('Etto... No puedo expulsar administradores (⌒_⌒;)', { reply_to_message_id: ctx.message!.message_id });
		return;
	}

	if (ctx.message!.reply_to_message.forward_from != undefined) {
		await ctx.kickChatMember(ctx.message!.reply_to_message.forward_from.id);
		await ctx.unbanChatMember(ctx.message!.reply_to_message.forward_from.id);
	} else {
		await ctx.kickChatMember(ctx.message!.reply_to_message.from!.id);
		await ctx.unbanChatMember(ctx.message!.reply_to_message.from!.id);
	}

	await ctx.replyWithChatAction('typing');
	await ctx.reply('¡Expulsad@!', { reply_to_message_id: ctx.message!.reply_to_message.message_id });
	await ctx.replyWithSticker('CAADAgADpxsAAuCjggf0_Fu4xLzgxAI');
});

Ban.command('kickme', async ctx => {
	let chatMember = await ctx.getChatMember(ctx.botInfo?.id ?? (await ctx.tg.getMe()).id);
	if (!chatMember.can_restrict_members) {
		await ctx.replyWithChatAction('typing');
		await ctx.reply('Etto... no tengo los permisos para expulsar usuarios (╥_╥)', {
			reply_to_message_id: ctx.message!.message_id
		});
		return;
	}

	chatMember = await ctx.getChatMember(ctx.message!.from!.id);
	if (chatMember.status == 'administrator' || chatMember.status == 'creator') {
		await ctx.replyWithChatAction('typing');
		await ctx.reply('Etto... No puedo expulsar administradores (⌒_⌒;)', { reply_to_message_id: ctx.message!.message_id });
		return;
	}

	await ctx.kickChatMember(ctx.message!.from!.id);
	await ctx.unbanChatMember(ctx.message!.from!.id);

	await ctx.replyWithChatAction('typing');
	await ctx.reply('¡Espero que vuelvas pronto! ( ; ω ; )', { reply_to_message_id: ctx.message!.message_id });
	await ctx.replyWithSticker('CAADAgADmhsAAuCjggeKLIclx5HvGQI');
});

Ban.command('unban', async ctx => {
	if (ctx.message!.reply_to_message == undefined) {
		await ctx.replyWithChatAction('typing');
		await ctx.reply('Uhm? No puedo desbanear a nadie si no respondes a un mensaje', {
			reply_to_message_id: ctx.message!.message_id
		});
		return;
	}

	let chatMember = await ctx.getChatMember(ctx.message!.from!.id);
	if (!chatMember.can_restrict_members && chatMember.status != 'creator') {
		await ctx.replyWithChatAction('typing');
		await ctx.reply('Lo siento, pero me dijeron que tú no puedes desbanear usuarios...', {
			reply_to_message_id: ctx.message!.message_id
		});
		return;
	}

	chatMember = await ctx.getChatMember(ctx.botInfo?.id ?? (await ctx.tg.getMe()).id);
	if (!chatMember.can_restrict_members) {
		await ctx.replyWithChatAction('typing');
		await ctx.reply('Etto... no tengo los permisos para desbanear usuarios (╥_╥)', {
			reply_to_message_id: ctx.message!.message_id
		});
		return;
	}

	if (ctx.message!.reply_to_message.forward_from != undefined) {
		ctx.unbanChatMember(ctx.message!.reply_to_message.forward_from.id);
	} else {
		ctx.unbanChatMember(ctx.message!.reply_to_message.from!.id);
	}

	await ctx.replyWithChatAction('typing');
	await ctx.reply('¡Desbanead@! (o^▽^o)', { reply_to_message_id: ctx.message!.reply_to_message.message_id });
	await ctx.replyWithSticker('CAADAgADshsAAuCjgge9W_YhLXZXrgI');
});

export default Ban;
